### ISSUE TITLE
Pipeline stuck in "Starting" after period of master unavailability

### ISSUE THREAD

**Issue by actgardner** [pachyderm/pachyderm#5298]
**What happened?**:

When running workloads automatically on a new Hub cluster in the examples runner, pipelines occasionally stick in a state where they are "starting" but can never make progress. My suspicion is that the GKE cluster begins to automatically resize (adding a nodepool) which makes the k8s master unavailable for several minutes. During this time we exhaust retries setting up the RCs.

In this case the examples runner got stuck running the openCV example: https://github.com/pachyderm/pachyderm/tree/master/examples/opencv

```
root@example-runner-1601056800-qqkt5:/app# pachctl list pipeline
NAME    VERSION INPUT                CREATED    STATE / LAST JOB    DESCRIPTION
montage 1       (edges:/ ⨯ images:/) 2 days ago starting / starting A pipeline that combines images from the `images` and `edges` repositories into a montage.
edges   1       images:/*            2 days ago starting / starting A pipeline that performs image edge detection by using the OpenCV library.
```

Logs:
```
2020-09-25T18:10:14Z ERROR PPS master: error retrieving RC for "montage": Get https://10.4.0.1:443/api/v1/namespaces/default/replicationcontrollers?labelSelector=pipelineName%3Dmontage: dial tcp 10.4.0.1:443: i/o timeout; retrying in 544.806308ms
2020-09-25T18:10:21Z ERROR PPS master: error retrieving RC for "montage": Get https://10.4.0.1:443/api/v1/namespaces/default/replicationcontrollers?labelSelector=pipelineName%3Dmontage: dial tcp 10.4.0.1:443: connect: connection refused; retrying in 758.900801ms
2020-09-25T18:10:22Z ERROR PPS master: Get https://10.4.0.1:443/api/v1/namespaces/default/replicationcontrollers?labelSelector=pipelineName%3Dmontage: dial tcp 10.4.0.1:443: connect: connection refused
```

No replication controllers have been created:
```
k get rc
No resources found in default namespace.
```

**What you expected to happen?**:

Pachd should be able to tolerate arbitrary outages of the k8s master and restart pipelines without intervention.

**How to reproduce it (as minimally and precisely as possible)?**:

Add a new nodepool to a GKE cluster and try to create a pipeline while the cluster is in maintenance mode. You can try to `kubectl` to confirm the master is inaccesible during the operation.

**Anything else we need to know?**:

**Environment?**:
- Kubernetes version (use `kubectl version`): N/A
- Pachyderm CLI and pachd server version (use `pachctl version`): `1.11.0`
- Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s): GKE
- OS (e.g. from /etc/os-release):
- Others:

**Comment by brycemcanally** [comment]
I've looked into this a little bit, and the issue does seem to be due to exhausting retries when attempting to fetch the RC. It looks like we are exhausting the retries here: https://github.com/pachyderm/pachyderm/blob/5fc5fae3ff7c35729102217cca46570503450b6b/src/server/pps/server/pipeline_controller.go#L235. Although this is in an infinite backoff, there is a max count for errors (3). It looks like we special case a particular set of errors that result with restarting the pipeline, but the rest will result with an error being returned. It also looks like we don't actually retry a step that errors: https://github.com/pachyderm/pachyderm/blob/5fc5fae3ff7c35729102217cca46570503450b6b/src/server/pps/server/master.go#L107 which I think means that we just lose the pipeline event? I guess this is because there are not very good recovery options here? @msteffen should be able to comment more on this, but I think we will probably need to make a change such that the k8s master being unavailable is considered a transient error by the retry logic. I think if this logic could be restructured such that the retries are based on an elapsed time with the exponential backoff  (15 mins is the default for our exponential backoff) rather than a count, we would be better off to handle transient errors. As of now, 3 retries is still in the <1s range for the backoffs. I don't think we want to be in the situation where we need to account for all of the potential transient errors.

**Comment by msteffen** [comment]
@brycemcanally's analysis makes sense to me.

I'll have to think a bit about some possible solutions, but one thing Bryce and I have discussed separately is having a background "polling" goro that checks the state of every pipeline at a regular interval and tries to adjust each pipeline's current state towards its target state (if it's not there already). I think such a mechanism would fix this, since the issue is the creation event being lost, and a background poller would eventually kick the STARTING pipeline and move it into running.

**Comment by msteffen** [comment]
This should be fixed by #5330 and #5387
